### PR TITLE
parameter dimensions in ArrayHelper::index

### PR DESCRIPTION
**Pull request by vlad-khramov**
Issue #10845

### CODE REVIEW

**Comment by SilverFire** [comment]
The idea is overally good, thank you.
Could you provide the benchmarking of the method before and after the changes (without dimensions).

Later `ActiveQuery::indexBy` might be adjusted to support this feature as well

**Comment by quantum13** [comment]
Is this benchmark enough?
```php
require(__DIR__ . '/../vendor/yiisoft/yii2/Yii.php');

$testArr = [];

for($i=0;$i<10000;$i++) {
    $testArr = ['id'=>$i, 'name'=>"name{$i}", 'data' => $i%5];
}

print "by id\n";
$start = microtime(true);
for($i=0;$i<10000;$i++) {
    \yii\helpers\ArrayHelper::index($testArr, 'id');
}
print microtime(true) - $start;
print "\n";

$start = microtime(true);
for($i=0;$i<10000;$i++) {
    \yii\helpers\ArrayHelper::index2($testArr, 'id');
}
print microtime(true) - $start;
print "\n";

print "by name\n";
$start = microtime(true);
for($i=0;$i<10000;$i++) {
    \yii\helpers\ArrayHelper::index($testArr, 'name');
}
print microtime(true) - $start;
print "\n";

$start = microtime(true);
for($i=0;$i<10000;$i++) {
    \yii\helpers\ArrayHelper::index2($testArr, 'name');
}
print microtime(true) - $start;
print "\n";

print "by data\n";
$start = microtime(true);
for($i=0;$i<10000;$i++) {
    \yii\helpers\ArrayHelper::index($testArr, 'data');
}
print microtime(true) - $start;
print "\n";

$start = microtime(true);
for($i=0;$i<10000;$i++) {
    \yii\helpers\ArrayHelper::index2($testArr, 'data');
}
print microtime(true) - $start;
print "\n";
```

```
by id
0.82908296585083
0.84308409690857
by name
0.77907800674438
0.83408308029175
by data
0.77907800674438
0.86208605766296
```

Without accepting string as third param (`if (!is_array($dimensions)) {`):
```
by id
0.80808091163635
0.85508489608765
by name
0.78507900238037
0.79908013343811
by data
0.77207684516907
0.81008100509644
```

**Comment by quantum13** [comment]
Is this benchmark enough?
```php

$testArr = [];

for($i=0;$i<1000;$i++) {
    $testArr[] = ['id'=>$i, 'name'=>"name{$i}", 'data' => $i%5];
}

print "by id\n";
$start = microtime(true);
for($i=0;$i<1000;$i++) {
    \yii\helpers\ArrayHelper::index($testArr, 'id');
}
print microtime(true) - $start;
print "\n";

$start = microtime(true);
for($i=0;$i<1000;$i++) {
    \yii\helpers\ArrayHelper::index2($testArr, 'id');
}
print microtime(true) - $start;
print "\n";

print "by name\n";
$start = microtime(true);
for($i=0;$i<1000;$i++) {
    \yii\helpers\ArrayHelper::index($testArr, 'name');
}
print microtime(true) - $start;
print "\n";

$start = microtime(true);
for($i=0;$i<1000;$i++) {
    \yii\helpers\ArrayHelper::index2($testArr, 'name');
}
print microtime(true) - $start;
print "\n";

print "by data\n";
$start = microtime(true);
for($i=0;$i<1000;$i++) {
    \yii\helpers\ArrayHelper::index($testArr, 'data');
}
print microtime(true) - $start;
print "\n";

$start = microtime(true);
for($i=0;$i<1000;$i++) {
    \yii\helpers\ArrayHelper::index2($testArr, 'data');
}
print microtime(true) - $start;
print "\n";
```

```
by id
16.528651952744
17.238723993301
by name
16.658666133881
17.389738798141
by data
16.49764919281
17.329732894897
```

**Comment by VirtualRJ** [comment]
Just a suggestion:

Why not instead of adding parameter `dimensions` add possibility to set array for the parameter `key` and add a boolean parameter `$includeIfNoKeyExits = false` (or a way better name than I can come up with right now). This way the scenario `$result = ArrayHelper::index($array, null, ['id']);` can become more readable and more consistent with current use (in my opinion ofcourse) `$result = ArrayHelper::index($array, ['id'], true);`. This way you keep the added functionality without the need for adding null values.

**Comment by quantum13** [comment]
@VirtualRJ yes, I thought about different variants. Adding only bool parameter can't help with two different common situations:
- index by one field (for example status) and add elements with the same status without keys.
```php
[
    'busy' => [
        [id = 1, status = busy],
        [id = 2, status = busy],
    ],
    'free' => [
        [id = 3, status = free],
        [id = 4, status = free],
    ],
]
````
- index by one field (status) and add elements with the same status with keys from enother field (for example id)
```php
[
    'busy' => [
        1 => [id = 1, status = busy],
        2 => [id = 2, status = busy],
    ],
    'free' => [
        3 => [id = 3, status = free],
        4 => [id = 4, status = free],
    ],
]
````
With suggested api it could be done as:
`ArrayHelper::index($array, null, 'status');` and `ArrayHelper::index($array, 'id', 'status');` 

---

As a variant api could be done in this way:
`ArrayHelper::index($array, ['status']);` and `ArrayHelper::index($array, ['status'], 'id');` 

It allow not to use nulls, but the second param changes meaning from key of last array to keys of dimensions.

**Comment by VirtualRJ** [comment]
@quantum13 I see. For your first example you might be able to say `['id','status']`, but your second one is not possible as far as I can see in my suggestion.

The variant api is I guess not an option because it will be BC breaking.

**Comment by SilverFire** [comment]
I'd like to merge this PR in 2.0.8, but the changes drop the performance. The algorithm looks over-complicated and might be simplified somehow.

File: framework/helpers/BaseArrayHelper.php [hunk 1]
```diff
@@ -247,7 +247,12 @@ public static function remove(&$array, $key, $default = null)
      * The key can be a key name of the sub-array, a property name of object, or an anonymous
      * function which returns the key value given an array element.
      *
-     * If a key value is null, the corresponding array element will be discarded and not put in the result.
+     * Dimensions is an array of keys, which will be used to produce multidimension result array.
```
Comment (samdark) [inline framework/helpers/BaseArrayHelper.php]: multidimension → multidimensional

File: framework/helpers/BaseArrayHelper.php [hunk 2]
```diff
@@ -247,7 +247,12 @@ public static function remove(&$array, $key, $default = null)
...
+     * This parameter is useful if indexing by non unique key value is needed.
+     *
+     * If key or a key value is null and dimensions are not specified, the corresponding array element
+     * will be discarded and not put in the result. Otherwise, if dimensions are specified, array element
```
Comment (samdark) [inline framework/helpers/BaseArrayHelper.php]: Remove "and not put in the result".

File: framework/helpers/BaseArrayHelper.php [hunk 3]
```diff
@@ -255,30 +260,80 @@ public static function remove(&$array, $key, $default = null)
...
      *     ['id' => '123', 'data' => 'abc'],
      *     ['id' => '345', 'data' => 'def'],
+     *     ['id' => '345', 'data' => 'hgi'],
      * ];
      * $result = ArrayHelper::index($array, 'id');
      * // the result is:
      * // [
      * //     '123' => ['id' => '123', 'data' => 'abc'],
-     * //     '345' => ['id' => '345', 'data' => 'def'],
+     * //     '345' => ['id' => '345', 'data' => 'hgi'], // will used last array element with the same key value
```
Comment (samdark) [inline framework/helpers/BaseArrayHelper.php]: will used → will use

File: framework/helpers/BaseArrayHelper.php [hunk 4]
```diff
@@ -255,30 +260,80 @@ public static function remove(&$array, $key, $default = null)
...
      * @param array $array the array that needs to be indexed
-     * @param string|\Closure $key the column name or anonymous function whose result will be used to index the array
+     * @param string|\Closure|null $key the column name or anonymous function whose result will be used to index the array
+     * @param array|string $dimensions
      * @return array the indexed array
      */
-    public static function index($array, $key)
+    public static function index($array, $key, $dimensions = [])
     {
+        if (!is_array($dimensions)) {
```
Comment (samdark) [inline framework/helpers/BaseArrayHelper.php]: Could be done by casting to array:

```php
$dimensions = (array) $dimensions;
```

File: framework/helpers/BaseArrayHelper.php [hunk 5]
```diff
@@ -255,30 +260,80 @@ public static function remove(&$array, $key, $default = null)
...
+            $dimensions = [$dimensions];
+        }
         $result = [];
         foreach ($array as $element) {
-            $value = static::getValue($element, $key);
-            $result[$value] = $element;
+            $lastArray = &$result;
+
+            foreach ($dimensions as $dimension) {
+                $val = static::getValue($element, $dimension);
```
Comment (samdark) [inline framework/helpers/BaseArrayHelper.php]: Better to use `$value` rather than `$val`.

**Comment by SilverFire** [comment]
Merged with adjustments: https://github.com/yiisoft/yii2/commit/f684bc955035c28f2b8e68d47c828328647085d3

**Comment by SilverFire** [comment]
Thank you!

**Pull request merged**